### PR TITLE
Contact JSBox resource needs to be able to update extras and subscriptions

### PR DESCRIPTION
**Pull request by justinvdm**

### CODE REVIEW

File: go/vumitools/contact/models.py [hunk 1]
```diff
@@ -130,8 +128,12 @@ def new_smart_group(self, name, query):
         yield group.save()
         returnValue(group)
 
+    @Manager.calls_manager
     def get_contact_by_key(self, key):
-        return self.contacts.load(key)
+        contact = yield self.contacts.load(key)
+        if contact is None:
+            raise RuntimeError("Contact with key '%s' not found." % key)
```
Comment (hodgestar) [inline go/vumitools/contact/models.py]: As a general rule, don't raise `RuntimeError` -- it makes catching this specific error hard. Make a specific exception class.

File: go/apps/jsbox/contacts.py [hunk 2]
```diff
@@ -88,9 +86,36 @@ def handle_update(self, api, command):
...
+    def _update_dynamic_field(self, field_name, api, command):
+        try:
+            if 'key' not in command:
+                raise SandboxError("'key' needs to be specified for command")
+
+            store = self._contact_store_for_api(api)
+            contact = yield store.get_contact_by_key(command['key'])
+
+            field = getattr(contact, field_name)
+            field.update(self.pick_fields(command, *field.keys()))
```
Comment (hodgestar) [inline go/apps/jsbox/contacts.py]: I think users need to be allowed to set keys that don't exist yet.

File: go/apps/jsbox/contacts.py [hunk 3]
```diff
@@ -88,9 +86,36 @@ def handle_update(self, api, command):
...
+            yield contact.save()
+        except (SandboxError, RuntimeError) as e:
```
Comment (hodgestar) [inline go/apps/jsbox/contacts.py]: This is why raising `RuntimeError` is bad -- this `except RuntimeError` clause could catch almost anything.

File: go/apps/jsbox/contacts.py [hunk 4]
```diff
@@ -63,12 +63,9 @@ def handle_get_or_create(self, api, command):
...
             contact=contact.get_data()))
 
-    def _filter_contact_fields(self, command):
-        """
-        Returns only the fields of a command that are Contact model fields.
-        """
-        return dict((k, command[k]) for k in Contact.field_descriptors.keys()
-                    if k in command)
+    @staticmethod
+    def pick_fields(collection, *fields):
```
Comment (hodgestar) [inline go/apps/jsbox/contacts.py:67]: Plus one on the new name. :)

File: go/apps/jsbox/contacts.py [hunk 5]
```diff
@@ -88,9 +90,42 @@ def handle_update(self, api, command):
...
+                raise SandboxError("'field' needs to be specified for command")
+
+            if 'value' not in command:
+                raise SandboxError("'value' needs to be specified for command")
+
+            store = self._contact_store_for_api(api)
+            contact = yield store.get_contact_by_key(command['key'])
+
+            field = getattr(contact, field_name)
+            field[command['field']] = command['value']
```
Comment (justinvdm) [inline go/apps/jsbox/contacts.py:102]: I'm not really happy about having the command look something like `{key: <contact_key>, field: <field_name>, value: <field_value>}`, any ideas on a better way to do this?

Comment (hodgestar) [inline go/apps/jsbox/contacts.py:102]: Maybe we should re-name key `contact_key` to differentiate it from `field` better?

File: go/apps/jsbox/contacts.py [hunk 6]
```diff
@@ -88,9 +90,42 @@ def handle_update(self, api, command):
...
+            yield contact.save()
+        except (SandboxError, ContactError) as e:
+            log.warning(str(e))
+            returnValue(self.reply(command, success=False, reason=unicode(e)))
+
+        returnValue(self.reply(
+            command,
+            success=True,
+            contact=contact.get_data()))
```
Comment (hodgestar) [inline go/apps/jsbox/contacts.py]: Should we be returning the contact here? I think just success is enough.

File: go/apps/jsbox/contacts.py [hunk 7]
```diff
@@ -63,12 +63,13 @@ def handle_get_or_create(self, api, command):
...
-        """
-        return dict((k, command[k]) for k in Contact.field_descriptors.keys()
-                    if k in command)
+    @staticmethod
+    def pick_fields(collection, *fields):
+        return dict((k, collection[k]) for k in fields if k in collection)
+
+    @staticmethod
+    def omit_fields(collection, *fields):
+        return dict((k, collection[k]) for k in fields if k not in collection)
```
Comment (hodgestar) [inline go/apps/jsbox/contacts.py]: Is anyone using omit_fields?

Comment (justinvdm) [inline go/apps/jsbox/contacts.py]: Ah, thanks, forgot to remove it

**Comment by hodgestar** [comment]
:+1:

**Pull request merged**